common/djangoapps/student/views.py

```diff
@@ -703,13 +703,34 @@ def dashboard(request):
     course_entitlement_available_sessions = {}
     for course_entitlement in course_entitlements:
         course_entitlement.update_expired_at()
-        # Filter only the course runs that do not have an enrollment_end date set, or have one set in the future
         course_runs_for_course = get_course_runs_for_course(str(course_entitlement.course_uuid))
         enrollable_course_runs = []
 
+        # Only show published course runs that can still be enrolled and upgraded
+        now = datetime.datetime.now(UTC)
         for course_run in course_runs_for_course:
+            # Only courses that have not ended will be displayed
+            run_start = course_run.get('start')
+            run_end = course_run.get('end')
+            is_running = start and (not run_end or (dateutil.parser.parse(run_end) > now))
+
+            # Only courses that can currently be enrolled in will be displayed
+            enrollment_start = course_run.get('enrollment_start')
             enrollment_end = course_run.get('enrollment_end')
-            if not enrollment_end or (dateutil.parser.parse(enrollment_end) > datetime.datetime.now(UTC)):
+            can_enroll = (not enrollment_start or dateutil.parser.parse(enrollment_start) < now
+                and (not enrollment_end or (dateutil.parser.parse(enrollment_end) > now)))
+
+            # Only upgrade-able courses will be displayed
+            upgrade_deadline = next(iter(
+                [seat.get('upgrade_deadline') for seat in course_run.get('seats')
+                 if seat.get('type') == course_entitlement.mode]), None
+            )
+            can_upgrade = not upgrade_deadline or (dateutil.parser.parse(upgrade_deadline) > datetime.datetime.now(UTC))
```

Comment:
can_upgrade = not upgrade_deadline or (dateutil.parser.parse(upgrade_deadline) > now